### PR TITLE
Introduction and implementation of Bookmark tool for: Qt Quick and Qt…

### PR DESCRIPTION
**Pull request by mfeigl**
Creating the Bookmark tools for the Qt Quick and Qt Widgets toolkit and Qt Qml.

@ldanzinger , may I please ask you to review the docs?

### CODE REVIEW

File: uitools/README.md
Comment (anmacdonald) [inline uitools/README.md:14]: 🙃 Extra tick!

File: uitools/cpp/Esri/ArcGISRuntime/Toolkit/BookmarkListItem.cpp
Comment (anmacdonald) [inline uitools/cpp/Esri/ArcGISRuntime/Toolkit/BookmarkListItem.cpp:42]: 🙃 Match order of arguments!

File: uitools/cpp/Esri/ArcGISRuntime/Toolkit/BookmarkListItem.h
Comment (anmacdonald) [inline uitools/cpp/Esri/ArcGISRuntime/Toolkit/BookmarkListItem.h:21]: 🔧 Looks like this can be removed.

File: uitools/cpp/Esri/ArcGISRuntime/Toolkit/BookmarksViewController.h
Comment (anmacdonald) [inline uitools/cpp/Esri/ArcGISRuntime/Toolkit/BookmarksViewController.h:26]: 🔧 Can be removed.

File: uitools/cpp/Esri/ArcGISRuntime/Toolkit/BookmarksViewController.h
Comment (anmacdonald) [inline uitools/cpp/Esri/ArcGISRuntime/Toolkit/BookmarksViewController.h:32]: 🔧 Can be removed, is implicitly included as part of `GenericListModel.h`.

File: uitools/examples/cpp_quick/src/demos/BookmarksViewDemo.cpp
Comment (anmacdonald) [inline uitools/examples/cpp_quick/src/demos/BookmarksViewDemo.cpp:13]: 🔧 Remove?

File: uitools/examples/cpp_quick/src/demos/BookmarksViewDemoForm.qml
Comment (anmacdonald) [inline uitools/examples/cpp_quick/src/demos/BookmarksViewDemoForm.qml:28]: 🙃 Place this line above anchors.

File: uitools/examples/qml_quick/src/demos/BookmarksViewDemoForm.qml
Comment (anmacdonald) [inline uitools/examples/qml_quick/src/demos/BookmarksViewDemoForm.qml:11]: This is entirely my fault. I think that even for the example app, that these headers are likely wrong. There is no sample code restrictions document to review in this repo. @ldanzinger, should the example app copyright headers be the Apache license ones?

File: uitools/examples/qml_quick/src/demos/BookmarksViewDemoForm.qml
Comment (anmacdonald) [inline uitools/examples/qml_quick/src/demos/BookmarksViewDemoForm.qml:36]: 🙃 Place this line above anchors.

File: uitools/import/Esri/ArcGISRuntime/Toolkit/BookmarksView.qml
Comment (anmacdonald) [inline uitools/import/Esri/ArcGISRuntime/Toolkit/BookmarksView.qml:32]: ❌  Copy+paste error here from BasemapGallery. The `Bookmarks` are associated with the `Geomodel`, so at most they can come from a Webmap/Webscene or be programmatically defined.

File: uitools/import/Esri/ArcGISRuntime/Toolkit/BookmarksView.qml
Comment (anmacdonald) [inline uitools/import/Esri/ArcGISRuntime/Toolkit/BookmarksView.qml:86]: 🔧 (This is personal preference and you are not beholden to follow through with this suggestion.)

I'm not sure I like the internal `Frame`, it leads to significant empty space about the margins. 

On reflection, I wonder if adding a frame/border should be the choice of the developer when they integrate the tool into their application, as opposed to being something we impose.

Of note: there's no border in the widgets variant.

Frame removed code:

```suggestion
    contentItem: ListView {
            id: listView
            model: controller.bookmarks
            clip: true

            ScrollBar.vertical: ScrollBar {
                active: true
                snapMode: ScrollBar.SnapOnRelease
            }

            delegate: ItemDelegate {
                // `name` is a role within GenericListModel's BookmarkListItem specification.
                // `text` can take `name` because the `model` builds the roles from the
                // properties of BookmarkListItem that is set in BookmarksViewController.
                text: name
                width: parent.width
                indicator: null
                onPressed: controller.zoomToBookmarkExtent(modelData)
            }
        }
    }
```

File: uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BookmarksViewController.qml
Comment (anmacdonald) [inline uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BookmarksViewController.qml:26]: 🙃 And the zoom functionality on the geoview!

File: uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BookmarksViewController.qml
Comment (anmacdonald) [inline uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BookmarksViewController.qml:42]: ❌  Not quite true! There's a `modelData` and a `name` property.

File: uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BookmarksViewController.qml
Comment (anmacdonald) [inline uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BookmarksViewController.qml:50]: ❌  Word salad. This is not for QML/C++ layer compatibility either. We take a `bookmark` and we zoom to that bookmark in the `GeoView`.

File: uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BookmarksViewController.qml
Comment (anmacdonald) [inline uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BookmarksViewController.qml:76]: 🙃 This is in the `internal` object so doesn't need documented like a a public property. Of note, it's worth documenting `rawBookmarks` as well and explain the rationale for both.

File: uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BookmarksViewController.qml
Comment (anmacdonald) [inline uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BookmarksViewController.qml:81]: 🙃 Quick comment for why we're doing this.

File: uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BookmarksViewController.qml
Comment (anmacdonald) [inline uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BookmarksViewController.qml:106]: 🙃 Quick comment what this is for.

File: uitools/widgets/Esri/ArcGISRuntime/Toolkit/BookmarksView.cpp
Comment (anmacdonald) [inline uitools/widgets/Esri/ArcGISRuntime/Toolkit/BookmarksView.cpp:42]: ❌ World salad, caused by BasemapGallery copy+paste. There's no such thing as developer bookmarks and bookmarks are only part of the Geomodel. 

When a bookmark is selected the geoview's viewpoint is set to the bookmark's stored viewpoint. Avoid mentioning rendering/replacement.

File: uitools/widgets/Esri/ArcGISRuntime/Toolkit/BookmarksView.cpp
Comment (anmacdonald) [inline uitools/widgets/Esri/ArcGISRuntime/Toolkit/BookmarksView.cpp:52]: ❌ No `setGeoModel` call here.

File: uitools/widgets/Esri/ArcGISRuntime/Toolkit/BookmarksView.cpp
Comment (anmacdonald) [inline uitools/widgets/Esri/ArcGISRuntime/Toolkit/BookmarksView.cpp:105]: 🙃 current bookmark AT index

File: uitools/widgets/Esri/ArcGISRuntime/Toolkit/BookmarksView.cpp
Comment (anmacdonald) [inline uitools/widgets/Esri/ArcGISRuntime/Toolkit/BookmarksView.cpp:107]: 🙃 When a click event is received, the bookmark at the given index is zoomed to via the controller.

File: uitools/widgets/Esri/ArcGISRuntime/Toolkit/BookmarksView.h
Comment (anmacdonald) [inline uitools/widgets/Esri/ArcGISRuntime/Toolkit/BookmarksView.h:48]: 🔧  Consider making a lambda within the constructor instead to hide implementation details from the public interface.

File: uitools/widgets/Esri/ArcGISRuntime/Toolkit/BookmarksView.h
Comment (anmacdonald) [inline uitools/widgets/Esri/ArcGISRuntime/Toolkit/BookmarksView.h:52]:
```suggestion
    Ui::BookmarksView* m_ui = nullptr;
```

**Review by anmacdonald — CHANGES_REQUESTED** [review]
👍  Very good work overall. 

Some tweaks needed to comments, mostly, but looks very solid.

**Comment by anmacdonald** [comment]
FYI, remember to update `CHANGELOG.md`!

**Review by lsmallwood — COMMENTED** [review]
Great work @mfeigl 👍

File: uitools/cpp/Esri/ArcGISRuntime/Toolkit/BookmarkListItem.cpp
Comment (mfeigl) [reply uitools/cpp/Esri/ArcGISRuntime/Toolkit/BookmarkListItem.cpp:66]: I believe the user has control over that. The tool never creates a `new` Bookmark, the tool just starts pointing to a whatever Bookmark the user has given. So wouldn't deleting the old bookmark here also pull the rug out from under the user if they still have the bookmark they've given to this tool either via the c'tor or this setter?

File: uitools/examples/qml_quick/src/demos/BookmarksViewDemoForm.qml
Comment (mfeigl) [reply uitools/examples/qml_quick/src/demos/BookmarksViewDemoForm.qml:11]: I do not know what copyright headers are required - I defer to the judgement of @ldanzinger

**Comment by mfeigl** [comment]
Thank you for the review - all resolved conversations have been accepted.

File: uitools/cpp/Esri/ArcGISRuntime/Toolkit/BookmarksViewController.cpp
Comment (anmacdonald) [reply uitools/cpp/Esri/ArcGISRuntime/Toolkit/BookmarksViewController.cpp]: Late to the party, but `connectToGeoView` guarantees the map and/or scene exists as it is only invoked once the geomodel is loaded.

**Review by anmacdonald — APPROVED** [review]
👍 LGTM, remember to get a review from Lucas for docs.

File: uitools/examples/qml_quick/src/demos/BookmarksViewDemoForm.qml
Comment (ldanzinger) [reply uitools/examples/qml_quick/src/demos/BookmarksViewDemoForm.qml:11]: Everything in this repo should be Apache license. @mfeigl can you open up a new issue and we can address it there please?

File: uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BookmarksViewController.qml
Comment (ldanzinger) [inline uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BookmarksViewController.qml:59]: assuming these are meant to be internal? They aren't enclosed in qdoc tags, just comments

File: uitools/widgets/Esri/ArcGISRuntime/Toolkit/BookmarksView.cpp
Comment (ldanzinger) [inline uitools/widgets/Esri/ArcGISRuntime/Toolkit/BookmarksView.cpp:39]:
```suggestion
  or are programmatically defined. When the user selects a bookmark from the provided list,
```

File: uitools/widgets/Esri/ArcGISRuntime/Toolkit/BookmarksView.cpp
Comment (ldanzinger) [inline uitools/widgets/Esri/ArcGISRuntime/Toolkit/BookmarksView.cpp:51]:
```suggestion
  The view maintains its associated controller, sets up the view itself and its model.
```

File: uitools/cpp/Esri/ArcGISRuntime/Toolkit/BookmarkListItem.cpp
Comment (mfeigl) [reply uitools/cpp/Esri/ArcGISRuntime/Toolkit/BookmarkListItem.cpp:71]: Removed the word `wrap` from this file, except from the class description so now it's in line with the rest of the toolkit documentation.

File: uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BookmarksViewController.qml
Comment (mfeigl) [reply uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BookmarksViewController.qml:59]: Yes, these are internal comments only - they are not to be published in the published documentation. They should, however, be present here in the code as comments. The formatting (i.e. wording) of these comments are the same as the docs, because that's a useful/easy-to-read formatting.

File: uitools/examples/cpp_quick/src/demos/BookmarksViewDemoForm.qml
Comment (mfeigl) [reply uitools/examples/cpp_quick/src/demos/BookmarksViewDemoForm.qml:28]: Yes, although some adjustments may be required. 
These `BookmarksViewDemo` files can be modified to load different maps (with or without bookmarks). Upon setting a new map, the new map will be displayed and the tool appends the list of bookmarks with the new bookmarks (nothing bad happens if the new map doesn't have bookmarks). 
However, the old bookmarks aren't removed. 

Should it be expected from the tool to remove the old bookmarks from the UI?

Comment (ldanzinger) [reply uitools/examples/cpp_quick/src/demos/BookmarksViewDemoForm.qml:28]: Closing out the loop - as discussed in scrum, it should remove the old bookmarks

**Review by lsmallwood — APPROVED** [review]
Looks great @mfeigl

**Pull request merged**
